[Rushil Ramesh] iP

rushilramesh opened the pull request:
# Duke Pro

As Kanye West once said,
> "We're living the future so
> the present is our past."

Duke Pro helps you plan your foreseeable future in a simple and efficient manner. It's
* text-based
* easy to learn
* ~~FAST~~ *SUPER FAST* to use

Can't wait to get started? Here's all you have to do!
1. Download it from [here](https://github.com/rushilramesh/ip/releases/tag/A-Jar).
2. double-click it.
3. add your tasks
4. let it manage your tasks for you 😉

The best part? It's 100% **FREE**!

Features:
- [x] Managing Tasks
- [x] Managing Events and Deadlines
- [ ] Reminders (coming soon)

---------------------
If you Java programmer, you can use it to practice Java too. Here's the `main` method:

    public static void main(String[] args) {
        new Duke(filePath).run();
    }

Commit:
Add Gradle support

rushilramesh committed:
Level-1

rushilramesh committed:
Level-2

rushilramesh committed:
Level-3

rushilramesh committed:
Level-4

rushilramesh committed:
A-TextUiTesting

rushilramesh committed:
Level-5

rushilramesh committed:
Level-6

rushilramesh committed:
branch-Level-7

rushilramesh committed:
Updated Duke.java for branch-Level-7

rushilramesh committed:
branch-Level-8

rushilramesh committed:
Merge branch 'branch-Level-8'

rushilramesh committed:
A-MoreOOP

rushilramesh committed:
A-Packages

rushilramesh committed:
A-Packages

rushilramesh committed:
A-JUnit

rushilramesh committed:
A-JavaDoc

rushilramesh committed:
A-CodingStandard

rushilramesh committed:
Level-9

rushilramesh committed:
Merge branch 'branch-A-CodingStandard'

rushilramesh committed:
Merge branch 'branch-Level-9'

rushilramesh committed:
Merge branch 'add-gradle-support'

rushilramesh committed:
Add GUI to Duke

skythefire commented on src/main/java/duke/task/Event.java:
```diff
@@ -0,0 +1,44 @@
+     * @param at date/time of event.
+     * @throws DateTimeParseException
+     */
+    public Event(String input, String at) throws DateTimeParseException{
+=======
+    public Event(String input, String at) throws DateTimeParseException {
```
Should the conflicts be merged before pushing to the master branch?

asherhy replied:
Agreed, conflicts should be resolved before committing and pushing

skythefire commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,59 @@
+import duke.ui.Ui;
+import duke.storage.Storage;
+
 public class Duke {
```
Perhaps group the relevant classes into packages?

asherhy replied:
Agree, would be better to group them and order them by alphabetical order.

skythefire commented on src/main/java/duke/commands/AddCommand.java:
```diff
@@ -0,0 +1,84 @@
+                    } else {
+                        Deadline deadline = new Deadline(details[0], details[1]);
+                        tasks.addToList(deadline);
+                        ui.printOutput("Got it. I've added this task:\n" + deadline + "\nNow you have " + tasks.getLength() + " tasks in the list.");
+                        storage.updateFile(tasks);
+                    }
+                } catch (DateTimeParseException e) {
+                    throw new DukeException("☹ OOPS!!! The date of the deadline is poorly formatted (d/MM/yyyy or d/MM/yyyy HHmm)");
+                }
+
+            } else {
+                throw new DukeException("☹ OOPS!!! The description of a deadline cannot be empty.");
+            }
+            break;
+        }
+        case EVENT: {
+            if (commands.length() > 0) {
+
+                try {
+                    String[] details = commands.split("/at ");
+                    if (details.length == 1) {
+                        throw new DukeException("☹ OOPS!!! Add a '/at time of event'");
+                    } else {
+                        Event event = new Event(details[0], details[1]);
+                        tasks.addToList(event);
+                        ui.printOutput("Got it. I've added this task:\n" + event + "\nNow you have " + tasks.getLength() + " tasks in the list.");
+                        storage.updateFile(tasks);
+                    }
+                } catch (DateTimeParseException e) {
+                    throw new DukeException("☹ OOPS!!! The date of the event is poorly formatted (d/MM/yyyy)");
+                }
+
+            } else {
+                throw new DukeException("☹ OOPS!!! The description of an event cannot be empty.");
+            }
+            break;
+        }
+        default: {
+            break;
+        }
```
I like your use of the switch statement here!

skythefire commented on src/main/java/duke/commands/FindCommand.java:
```diff
@@ -0,0 +1,43 @@
+package duke.commands;
```
I like how you packaged the relevant classes into sub-packages! :)

skythefire commented on src/main/java/duke/task/TaskList.java:
```diff
@@ -0,0 +1,107 @@
+     *
+     * @param date Date of tasks to be searched for
+     * @return array of tasks that are on the same date as the inputted date.
+     */
+
+    public Task[] tasksOnDate(String date) throws DukeException {
```
Perhaps remove the empty line between the JavaDoc and the method name?

skythefire reviewed (commented):
LGTM! Great work on iP 😃 All the best for the future Levels too!

rushilramesh committed:
Update Event.java

rushilramesh committed:
Add GUI to Duke

rushilramesh committed:
Add checkstyle funcitonality

asherhy commented on src/main/java/duke/commands/NonsenseCommand.java:
```diff
@@ -0,0 +1,17 @@
+public class NonsenseCommand extends Command{
+    // empty as the string input is invalid
+    public NonsenseCommand() {
+
```
possible to be removed if not needed

asherhy commented on src/main/java/duke/task/Deadline.java:
```diff
@@ -0,0 +1,40 @@
+    }
+
+    @Override
+    public boolean compareDate(LocalDate compare) {
```
boolean should change to isDateEqual or something similar

asherhy reviewed (commented):
Overall code is relatively clean and simple. Good job!

frederickpek commented on src/main/java/duke/commands/DeleteCommand.java:
```diff
@@ -0,0 +1,32 @@
+    }
+
+    @Override
+    public String execute(TaskList tasks, Storage storage) throws DukeException {
```
For your execute method here, you may want to "make the happy path more prominant" under the code quality section of the SE textbook. This should allow others to understand the exceptions are you checking for much easily and allow them to understand the main code easily.

frederickpek commented on src/main/java/duke/commands/UpdateCommand.java:
```diff
@@ -0,0 +1,32 @@
+    }
+
+    @Override
+    public String execute(TaskList tasks, Storage storage) throws DukeException {
```
The same thing applies here, to "make the happy path prominant".

frederickpek commented on src/main/java/duke/ui/Ui.java:
```diff
@@ -0,0 +1,65 @@
+    /**
+     * Prints welcome message when Duke is run.
+     */
+    public void showWelcome() {
```
I like that you use verbs to name your methods! Perhaps your method name could be named printWelcome instead. This should align more with the description of your method.

frederickpek commented on src/main/java/duke/ui/Ui.java:
```diff
@@ -0,0 +1,65 @@
+     *
+     * @param error error message to be printed.
+     */
+    public void showError(String error) {
```
Same here, perhaps 'printError' would be better.

frederickpek commented on src/main/java/duke/ui/Ui.java:
```diff
@@ -0,0 +1,65 @@
+    /**
+     * Prints a line to indicate the start and end of the output.
+     */
+    public void showLine() {
```
Same here, perhaps 'printLine' would be better.

frederickpek commented on src/main/java/duke/ui/Ui.java:
```diff
@@ -0,0 +1,65 @@
+    /**
+     * Prints error message if the filepath could not be found.
+     */
+    public void showLoadingError() {
```
Same here, 'printLoadingError' would be better.

frederickpek reviewed (commented):
Looks good! Just some code quality issues for you to look over 👍🏻

rushilramesh committed:
Classes do not have any assertions to check for
assumptions made.

Failing to check for these invariants or assumptions made could result
in bugs that could cause the program
to crash.

Add assertions to check for and document assumptions that should hold
up at various points in the code.

Assertions help check the state of the program at a particular point in
the code and can be enabled or disabled accordingly.

rushilramesh committed:
Add Assertions
Classes do not have any assertions to check for
assumptions made.

Failing to check for these invariants or assumptions made could result
in bugs that could cause the program
to crash.

Add assertions to check for and document assumptions that should hold
up at various points in the code.

Assertions help check the state of the program at a particular point in
the code and can be enabled or disabled accordingly.

rushilramesh committed:
Improve code quality
Overall code quality can be improved.

Poor code quality reduces the readablity of the code and makes it
harder for other programmers to understand the code.

Adhere to CS2103T's coding standards to standardise the code.

This improves the overall quality fo the code and makes it easier to
understand.

rushilramesh committed:
Complete A-CodeQuality
Overall code quality can be improved.

Poorly written code reduces the readablity of the code and makes it
harder for other programmers to understand.

Adhering to the CS2103 coding standards to standardise the way the code
is written.

This helps other programmers easily understand the code.

rushilramesh committed:
Merge branch 'branch-A-CodeQuality'

rushilramesh committed:
Add stream functionality

rushilramesh committed:
Add multiple delete feature
User does not have the option to delete multiple objects in one go.

User would manually have to delete each task individually in order to
delete multiple items.

Add multiple delete feature that allows user to input various indexes
separated by a whitespace to delete the tasks at the given positions in
the list.

This would streamline the deleting process for the user.

rushilramesh committed:
Add Better Search feature
Current search feature can only find a task if there is a complete
match.

This prevents the user from easily and quickly searching for tasks as
the user would be required to type the entire task name.

The better search allows the user to type a keyword that paritally
matches the task title.

This is a more efficient way of searching for tasks.

rushilramesh committed:
Merge pull request #1 from rushilramesh/branch-C-BetterSearch

Add Better Search feature

rushilramesh committed:
Add message to Assertions

rushilramesh committed:
Add assertions

rushilramesh committed:
Add Assertions functionality

rushilramesh committed:
Merge branch 'master' of https://github.com/rushilramesh/ip

rushilramesh committed:
Modift assertions

rushilramesh committed:
Implement better GUI

rushilramesh committed:
Update user guide

rushilramesh committed:
Update Ui.png

rushilramesh committed:
Update empty tasklist message

rushilramesh committed:
Improve code quality
